### PR TITLE
EpgContainer: use shared_ptr to manage CEpg pointers

### PR DESCRIPTION
**Pull request by MaxKellermann**

### CODE REVIEW

File: xbmc/epg/EpgContainer.cpp [hunk 1]
```diff
@@ -334,13 +333,13 @@ void CEpgContainer::Process(void)
   }
 }
 
-CEpg *CEpgContainer::GetById(int iEpgId) const
+std::shared_ptr<CEpg> CEpgContainer::GetById(int iEpgId) const
```
Comment (Jalle19) [inline xbmc/epg/EpgContainer.cpp]: We use typedefs for situations like these elsewhere in the code, i.e. `CEpgPtr`. Can you add that?

File: xbmc/epg/EpgContainer.cpp [hunk 2]
```diff
@@ -245,12 +244,12 @@ bool CEpgContainer::PersistAll(void)
...
   bool bReturn(true);
   m_critSection.lock();
-  std::map<unsigned int, CEpg*> copy = m_epgs;
+  const auto copy = m_epgs;
   m_critSection.unlock();
 
   for (EPGMAP_CITR it = copy.begin(); it != copy.end() && !m_bStop; it++)
   {
-    CEpg *epg = it->second;
+    CEpg *epg = it->second.get();
```
Comment (Jalle19) [inline xbmc/epg/EpgContainer.cpp]: No need to use `get()` here if you use `CEpgPtr epg = it->second;`.

File: xbmc/epg/EpgContainer.h [hunk 3]
```diff
@@ -281,7 +283,7 @@ namespace EPG
 
     void InsertFromDatabase(int iEpgID, const std::string &strName, const std::string &strScraperName);
 
-    typedef std::map<unsigned int, CEpg*> EPGMAP;
+    typedef std::map<unsigned int, std::shared_ptr<CEpg>> EPGMAP;
     typedef EPGMAP::iterator              EPGMAP_ITR;
```
Comment (Jalle19) [inline xbmc/epg/EpgContainer.h]: In a slightly unrelated note these could be removed in favor of using `auto` at some point.

**Comment by Jalle19** [comment]
@MaxKellermann does this fix any particular bug for you? Not saying it's a bad idea but it will probably require a lot of testing.

Ping @ksooo @xhaggi

**Comment by MaxKellermann** [comment]
Yes, it fixes this crash:

Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f9fbed89293 in std::local_Rb_tree_increment (__x=0x12) at ../../../../../src/libstdc++-v3/src/c++98/tree.cc:65
65      ../../../../../src/libstdc++-v3/src/c++98/tree.cc: No such file or directory.
[Current thread is 1 (Thread 0x7f9f61ffb700 (LWP 5694))]
(gdb) bt
#0  0x00007f9fbed89293 in std::local_Rb_tree_increment (__x=0x12) at ../../../../../src/libstdc++-v3/src/c++98/tree.cc:65
#1  std::_Rb_tree_increment (__x=0x7f9f681bebc0) at ../../../../../src/libstdc++-v3/src/c++98/tree.cc:91
#2  0x0000000000a1794c in std::_Rb_tree_iterator<std::pair<int const, std::shared_ptr<EPG::CEpgInfoTag> > >::operator++ (
    this=<synthetic pointer>) at /usr/include/c++/5/bits/stl_tree.h:205
#3  EPG::CEpg::Persist (this=this@entry=0x7f9f782ed8c0) at Epg.cpp:543
#4  0x0000000000a1e07d in EPG::CEpgContainer::PersistAll (this=this@entry=0x16b8360 <EPG::CEpgContainer::Get()::epgInstance>)
    at EpgContainer.cpp:256
#5  0x0000000000a1e488 in EPG::CEpgContainer::Process (this=0x16b8360 <EPG::CEpgContainer::Get()::epgInstance>) at EpgContainer.cpp:329
#6  0x000000000113f56f in CThread::Action (this=0x16b8410 <EPG::CEpgContainer::Get()::epgInstance+176>) at Thread.cpp:221
#7  0x000000000113f832 in CThread::staticThread (data=0x16b8410 <EPG::CEpgContainer::Get()::epgInstance+176>) at Thread.cpp:131
#8  0x00007f9fc6e570a4 in start_thread (arg=0x7f9f61ffb700) at pthread_create.c:309
#9  0x00007f9fbe80c06d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) t 2
[Switching to thread 2 (Thread 0x7f9f81a46700 (LWP 5663))]
#0  __lll_lock_wait () at ../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
135     ../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S: No such file or directory.
(gdb) bt
#0  __lll_lock_wait () at ../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
#1  0x00007f9fc6e594d4 in _L_lock_986 () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x00007f9fc6e59336 in __GI___pthread_mutex_lock (mutex=0x7f9f78095728) at ../nptl/pthread_mutex_lock.c:114
#3  0x00007f9fc49dbdc5 in sqlite3_exec () from /usr/lib/x86_64-linux-gnu/libsqlite3.so.0
#4  0x00000000009da947 in dbiplus::SqliteDataset::exec (this=0x7f9f780553c0, sql=...) at sqlitedataset.cpp:623
#5  0x00000000009c7f74 in CDatabase::ExecuteQuery (this=this@entry=0x16b86b0 <EPG::CEpgContainer::Get()::epgInstance+848>, 
    strQuery="DELETE FROM epg  WHERE idEpg = 905") at Database.cpp:224
#6  0x00000000009c811f in CDatabase::DeleteValues (this=this@entry=0x16b86b0 <EPG::CEpgContainer::Get()::epgInstance+848>, 
    strTable="epg", filter=...) at Database.cpp:185
#7  0x0000000000a21c9f in EPG::CEpgDatabase::Delete (this=this@entry=0x16b86b0 <EPG::CEpgContainer::Get()::epgInstance+848>, table=...)
    at EpgDatabase.cpp:143
#8  0x0000000000a1b70c in EPG::CEpgContainer::DeleteEpg (this=0x16b8360 <EPG::CEpgContainer::Get()::epgInstance>, epg=..., 
    bDeleteFromDatabase=<optimized out>) at EpgContainer.cpp:470
#9  0x0000000000c245a2 in PVR::CPVRChannel::Delete (this=this@entry=0x7f9f781d9380) at PVRChannel.cpp:153
#10 0x0000000000c27a36 in PVR::CPVRChannelGroup::RemoveDeletedChannels (this=this@entry=0x7f9f7800de90, channels=...)
    at PVRChannelGroup.cpp:672
#11 0x0000000000c299e3 in PVR::CPVRChannelGroup::UpdateGroupEntries (this=this@entry=0x7f9f7800de90, channels=...)
    at PVRChannelGroup.cpp:716
#12 0x0000000000c320a9 in PVR::CPVRChannelGroupInternal::UpdateGroupEntries (this=0x7f9f7800de90, channels=...)
    at PVRChannelGroupInternal.cpp:295
#13 0x0000000000c30f5b in PVR::CPVRChannelGroupInternal::Update (this=0x7f9f7800de90) at PVRChannelGroupInternal.cpp:126
#14 0x0000000000c32b39 in PVR::CPVRChannelGroups::Update (this=0x7f9f7c0008c0, bChannelsOnly=bChannelsOnly@entry=true)
    at PVRChannelGroups.cpp:215
#15 0x0000000000c3795c in PVR::CPVRChannelGroupsContainer::Update (this=0x7f9f7c000c40, bChannelsOnly=<optimized out>)
    at PVRChannelGroupsContainer.cpp:56
#16 0x0000000000c50331 in PVR::CPVRManager::ExecutePendingJobs (this=this@entry=0x16c2b20 <PVR::CPVRManager::Get()::pvrManagerInstance>)
    at PVRManager.cpp:1670
#17 0x0000000000c55c18 in PVR::CPVRManager::Process (this=0x16c2b20 <PVR::CPVRManager::Get()::pvrManagerInstance>) at PVRManager.cpp:530
#18 0x000000000113f56f in CThread::Action (this=0x16c2b28 <PVR::CPVRManager::Get()::pvrManagerInstance+8>) at Thread.cpp:221
#19 0x000000000113f832 in CThread::staticThread (data=0x16c2b28 <PVR::CPVRManager::Get()::pvrManagerInstance+8>) at Thread.cpp:131
#20 0x00007f9fc6e570a4 in start_thread (arg=0x7f9f81a46700) at pthread_create.c:309
#21 0x00007f9fbe80c06d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111

**Comment by fritsch** [comment]
Not saying the code is wrong, but one needs to understand _why_ the crash
happened within current design and see if using shared_ptrs would just
hides (!) the issue or if it really solves the issue.

2015-10-17 8:47 GMT+02:00 Max Kellermann <notifications@github.com>:

> Yes, it fixes this crash:
>
> Program terminated with signal SIGSEGV, Segmentation fault.
> #0 0x00007f9fbed89293 in std::local_Rb_tree_increment (__x=0x12) at
> ../../../../../src/libstdc++-v3/src/c++98/tree.cc:65
> 65 ../../../../../src/libstdc++-v3/src/c++98/tree.cc: No such file or
> directory.
> Current thread is 1 (Thread 0x7f9f61ffb700 (LWP 5694)) <http://gdb> bt
> #0 0x00007f9fbed89293 in std::local_Rb_tree_increment (__x=0x12) at
> ../../../../../src/libstdc++-v3/src/c++98/tree.cc:65
> #1 <https://github.com/xbmc/xbmc/pull/1> std::_Rb_tree_increment
> (__x=0x7f9f681bebc0) at ../../../../../src/libstdc++-v3/src/c++98/tree.cc:91
> #2 <https://github.com/xbmc/xbmc/pull/2> 0x0000000000a1794c in
> std::_Rb_tree_iterator > >::operator++ (
> this=) at /usr/include/c++/5/bits/stl_tree.h:205
> #3 <https://github.com/xbmc/xbmc/pull/3> EPG::CEpg::Persist
> (this=this@entry=0x7f9f782ed8c0) at Epg.cpp:543
> #4 <https://github.com/xbmc/xbmc/pull/4> 0x0000000000a1e07d in
> EPG::CEpgContainer::PersistAll (this=this@entry=0x16b8360
> EPG::CEpgContainer::Get()::epgInstance)
> at EpgContainer.cpp:256
> #5 <https://github.com/xbmc/xbmc/pull/5> 0x0000000000a1e488 in
> EPG::CEpgContainer::Process (this=0x16b8360
> EPG::CEpgContainer::Get()::epgInstance) at EpgContainer.cpp:329
> #6 <https://github.com/xbmc/xbmc/pull/6> 0x000000000113f56f in
> CThread::Action (this=0x16b8410 EPG::CEpgContainer::Get()::epgInstance+176)
> at Thread.cpp:221
> #7 <https://github.com/xbmc/xbmc/pull/7> 0x000000000113f832 in
> CThread::staticThread (data=0x16b8410
> EPG::CEpgContainer::Get()::epgInstance+176) at Thread.cpp:131
> #8 <https://github.com/xbmc/xbmc/pull/8> 0x00007f9fc6e570a4 in
> start_thread (arg=0x7f9f61ffb700) at pthread_create.c:309
> #9 <https://github.com/xbmc/xbmc/pull/9> 0x00007f9fbe80c06d in clone ()
> at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
> (gdb) t 2
> [Switching to thread 2 (Thread 0x7f9f81a46700 (LWP 5663))]
> #0 __lll_lock_wait () at
> ../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
> 135 ../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S: No such file or
> directory.
> (gdb) bt
> #0 __lll_lock_wait () at
> ../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
> #1 <https://github.com/xbmc/xbmc/pull/1> 0x00007f9fc6e594d4 in
> _L_lock_986 () from /lib/x86_64-linux-gnu/libpthread.so.0
> #2 <https://github.com/xbmc/xbmc/pull/2> 0x00007f9fc6e59336 in
> __GI___pthread_mutex_lock (mutex=0x7f9f78095728) at
> ../nptl/pthread_mutex_lock.c:114
> #3 <https://github.com/xbmc/xbmc/pull/3> 0x00007f9fc49dbdc5 in
> sqlite3_exec () from /usr/lib/x86_64-linux-gnu/libsqlite3.so.0
> #4 <https://github.com/xbmc/xbmc/pull/4> 0x00000000009da947 in
> dbiplus::SqliteDataset::exec (this=0x7f9f780553c0, sql=...) at
> sqlitedataset.cpp:623
> #5 <https://github.com/xbmc/xbmc/pull/5> 0x00000000009c7f74 in
> CDatabase::ExecuteQuery (this=this@entry=0x16b86b0
> EPG::CEpgContainer::Get()::epgInstance+848,
> strQuery="DELETE FROM epg WHERE idEpg = 905") at Database.cpp:224
> #6 <https://github.com/xbmc/xbmc/pull/6> 0x00000000009c811f in
> CDatabase::DeleteValues (this=this@entry=0x16b86b0
> EPG::CEpgContainer::Get()::epgInstance+848,
> strTable="epg", filter=...) at Database.cpp:185
> #7 <https://github.com/xbmc/xbmc/pull/7> 0x0000000000a21c9f in
> EPG::CEpgDatabase::Delete (this=this@entry=0x16b86b0
> EPG::CEpgContainer::Get()::epgInstance+848, table=...)
> at EpgDatabase.cpp:143
> #8 <https://github.com/xbmc/xbmc/pull/8> 0x0000000000a1b70c in
> EPG::CEpgContainer::DeleteEpg (this=0x16b8360
> EPG::CEpgContainer::Get()::epgInstance, epg=...,
> bDeleteFromDatabase=) at EpgContainer.cpp:470
> #9 <https://github.com/xbmc/xbmc/pull/9> 0x0000000000c245a2 in
> PVR::CPVRChannel::Delete (this=this@entry=0x7f9f781d9380) at
> PVRChannel.cpp:153
> #10 <https://github.com/xbmc/xbmc/pull/10> 0x0000000000c27a36 in
> PVR::CPVRChannelGroup::RemoveDeletedChannels (this=this@entry=0x7f9f7800de90,
> channels=...)
> at PVRChannelGroup.cpp:672
> #11 <https://github.com/xbmc/xbmc/pull/11> 0x0000000000c299e3 in
> PVR::CPVRChannelGroup::UpdateGroupEntries (this=this@entry=0x7f9f7800de90,
> channels=...)
> at PVRChannelGroup.cpp:716
> #12 <https://github.com/xbmc/xbmc/pull/12> 0x0000000000c320a9 in
> PVR::CPVRChannelGroupInternal::UpdateGroupEntries (this=0x7f9f7800de90,
> channels=...)
> at PVRChannelGroupInternal.cpp:295
> #13 <https://github.com/xbmc/xbmc/pull/13> 0x0000000000c30f5b in
> PVR::CPVRChannelGroupInternal::Update (this=0x7f9f7800de90) at
> PVRChannelGroupInternal.cpp:126
> #14 <https://github.com/xbmc/xbmc/pull/14> 0x0000000000c32b39 in
> PVR::CPVRChannelGroups::Update (this=0x7f9f7c0008c0,
> bChannelsOnly=bChannelsOnly@entry=true)
> at PVRChannelGroups.cpp:215
> #15 <https://github.com/xbmc/xbmc/pull/15> 0x0000000000c3795c in
> PVR::CPVRChannelGroupsContainer::Update (this=0x7f9f7c000c40,
> bChannelsOnly=)
> at PVRChannelGroupsContainer.cpp:56
> #16 <https://github.com/xbmc/xbmc/pull/16> 0x0000000000c50331 in
> PVR::CPVRManager::ExecutePendingJobs (this=this@entry=0x16c2b20
> PVR::CPVRManager::Get()::pvrManagerInstance)
> at PVRManager.cpp:1670
> #17 <https://github.com/xbmc/xbmc/pull/17> 0x0000000000c55c18 in
> PVR::CPVRManager::Process (this=0x16c2b20
> PVR::CPVRManager::Get()::pvrManagerInstance) at PVRManager.cpp:530
> #18 <https://github.com/xbmc/xbmc/pull/18> 0x000000000113f56f in
> CThread::Action (this=0x16c2b28
> PVR::CPVRManager::Get()::pvrManagerInstance+8) at Thread.cpp:221
> #19 <https://github.com/xbmc/xbmc/pull/19> 0x000000000113f832 in
> CThread::staticThread (data=0x16c2b28
> PVR::CPVRManager::Get()::pvrManagerInstance+8) at Thread.cpp:131
> #20 <https://github.com/xbmc/xbmc/pull/20> 0x00007f9fc6e570a4 in
> start_thread (arg=0x7f9f81a46700) at pthread_create.c:309
> #21 <https://github.com/xbmc/xbmc/pull/21> 0x00007f9fbe80c06d in clone ()
> at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8238#issuecomment-148891347>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by MaxKellermann** [comment]
The crash reason is easy to see once you understand how locking works:

CEpgContainer::m_critSection protectes the std::map m_epgs which contains CEpg pointers. All access to m_epgs must be protected with m_critSection. What the authors of that code didn't consider: if a CEpg pointer escapes that protected code section, it can be invalidated at any time. Which implies: it is illegal to leak CEpg pointers out of CEpgContainer::m_critSection protection!

That, however, is what commit 2ba1c43731c482661f61c7009f99dd089ddfdb6d implements, and what has been done wrong many times before, for example GetById() and GetByChannel() leak those pointers, explicitly. All of that is wrong, and each one of them can crash Kodi.

So what my change does is ensure that all CEpg instances remain valid as long as somebody is still holding a pointer, by wrapping all CEpg pointers in std::shared_ptr.

Understood?

**Comment by fritsch** [comment]
I only wanted to make sure, that it is done right once and for all. No "short fixes" should be implemented without understanding the issue in detail, while knowing the mess that is done here over the years.

So, then you have qualified, thank you :-) and nice that you cared to fix this mess.

jenkins build this please

PS: No need to explain me how locking works - I am into the kodi mess for > 5 years.

**Comment by MaxKellermann** [comment]
I'll update my commit message, and I'll resubmit with the CEpgPtr typedef. Don't merge just yet.

**Comment by fritsch** [comment]
Don't force push for the next 20 minutes please ... so that jenkins can make a first check, please.

**Comment by MaxKellermann** [comment]
Ok, until jenkins is done, I have pushed an updated commit in branch https://github.com/MaxKellermann/xbmc/commits/epg_shared_ptr2

Changes:
- unrelated build failure fix removed
- typedef CEpgPtr added
- shared_ptr.get() removed, replaced with "auto &epg =" ..., note the reference which avoids overhead by not increasing the reference counter unnecessarily; OTOH it adds another indirection; both variants are worse than my initial code (at runtime) - you decide what you want to have!
- more information in commit message

Tell me when I can overwrite branch epg_shared_ptr to update this pull request.

**Comment by fritsch** [comment]
Just force push into your original branch (after jenkins is done).

**Comment by MaxKellermann** [comment]
Pushed updated commit into this branch.

**Comment by Jalle19** [comment]
jenkins build this please again

**Comment by MartijnKaijser** [comment]
Hold on. Why is this against Isengard? This is not acceptable

**Comment by fritsch** [comment]
Good catch. @MaxKellermann please rebase onto current master.

This change and fixup is wanted - but Isengard is closed for such big rewrites.

**Pull request closed without merging**